funsor/interpreter.py

```diff
@@ -324,13 +333,54 @@ def dispatched_interpretation(fn):
     return fn
 
 
+class StatefulInterpretationMeta(type):
```

Comment:
I guess this is required for Python 3.5?  Once we drop Python 3.5 support, we should remove this and other metaclasses where possible in favor of `__init_subclass__`.

Reply:
Agreed, I was almost tempted to drop support for Python 3.5 in this PR, but that seemed like overkill.